relayer/core/relay.go

```diff
@@ -1,222 +1,66 @@
-// Copyright 2020 Snowfork
-// SPDX-License-Identifier: LGPL-3.0-only
-
 package core
 
 import (
-	"context"
-	"errors"
-	"fmt"
-	"os"
-	"os/signal"
-	"strings"
-	"syscall"
-	"time"
-
-	log "github.com/sirupsen/logrus"
-	"github.com/spf13/viper"
-	"golang.org/x/sync/errgroup"
+	"github.com/sirupsen/logrus"
 
-	"github.com/snowfork/polkadot-ethereum/relayer/chain/ethereum"
-	"github.com/snowfork/polkadot-ethereum/relayer/chain/parachain"
-	"github.com/snowfork/polkadot-ethereum/relayer/chain/relaychain"
 	"github.com/snowfork/polkadot-ethereum/relayer/workers"
 	"github.com/snowfork/polkadot-ethereum/relayer/workers/beefyrelayer"
-	"github.com/snowfork/polkadot-ethereum/relayer/workers/beefyrelayer/store"
 	"github.com/snowfork/polkadot-ethereum/relayer/workers/ethrelayer"
 	"github.com/snowfork/polkadot-ethereum/relayer/workers/parachaincommitmentrelayer"
 )
 
-type Relay struct {
-	parachainCommitmentRelayer *parachaincommitmentrelayer.Worker
-	beefyRelayer               *beefyrelayer.Worker
-	ethRelayer                 *ethrelayer.Worker
-}
-
-type WorkerConfig struct {
-	ParachainCommitmentRelayer workers.WorkerConfig `mapstructure:"parachaincommitmentrelayer"`
-	BeefyRelayer               workers.WorkerConfig `mapstructure:"beefyrelayer"`
-	EthRelayer                 workers.WorkerConfig `mapstructure:"ethrelayer"`
-}
-
-type Config struct {
-	Eth                  ethereum.Config   `mapstructure:"ethereum"`
-	Parachain            parachain.Config  `mapstructure:"parachain"`
-	Relaychain           relaychain.Config `mapstructure:"relaychain"`
-	BeefyRelayerDatabase store.Config      `mapstructure:"database"`
-	Workers              WorkerConfig      `mapstructure:"workers"`
-}
+type Relay struct{}
 
-func NewRelay() (*Relay, error) {
+func (re *Relay) Run() error {
 	config, err := LoadConfig()
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	var parachainCommitmentRelayer *parachaincommitmentrelayer.Worker
-
-	if config.Workers.ParachainCommitmentRelayer.Enabled {
-		parachainCommitmentRelayer, err = parachaincommitmentrelayer.NewWorker(&config.Parachain, &config.Relaychain, &config.Eth)
-		if err != nil {
-			return nil, err
-		}
-	}
-
-	var beefyRelayer *beefyrelayer.Worker
-
-	if config.Workers.BeefyRelayer.Enabled {
-		beefyRelayer, err = beefyrelayer.NewWorker(&config.Relaychain, &config.Eth, &config.BeefyRelayerDatabase)
-		if err != nil {
-			return nil, err
-		}
-	}
-
-	var ethRelayer *ethrelayer.Worker
+	var pool workers.WorkerPool
 
 	if config.Workers.EthRelayer.Enabled {
```

Comment:
The `Enabled` checks are duplicative. The worker pool will discard workers that are disabled [here](https://github.com/Snowfork/polkadot-ethereum/blob/ceea5878f722f23906521ba39cdb8aa9f07c63dc/relayer/workers/worker.go#L120-L122). So, if you prefer, you can instantiate the pool without appending like:
```
pool := workers.WorkerPool{
    ethrelayerFactory,
    beefyrelayerFactory,
    parachaincommitmentrelayerFactory,
}
```